dhcpv4/bsdp.go

```diff
@@ -0,0 +1,328 @@
+		bytes[0] |= 0x80
+	}
+	bytes[0] |= b.imageKind
+
+	// Index
+	binary.BigEndian.PutUint16(bytes[2:], b.index)
+	return
+}
+
+// BootImageIDFromBytes deserializes a collection of 4 bytes to a BootImageID.
+func bootImageIDFromBytes(bytes []byte) BootImageID {
+	return BootImageID{
+		isInstall: bytes[0]&0x80 != 0,
+		imageKind: bytes[0] & 0x7f,
+		index:     binary.BigEndian.Uint16(bytes[2:]),
+	}
+}
+
+// BootImage describes a boot image - contains the boot image ID and the name.
+type BootImage struct {
+	ID BootImageID
+	// This is a utf-8 string.
+	Name string
+}
+
+// toBytes converts a BootImage to a slice of bytes.
+func (b *BootImage) toBytes() (bytes []byte) {
+	idBytes := b.ID.toBytes()
+	bytes = append(bytes, idBytes[:]...)
+	bytes = append(bytes, byte(len(b.Name)))
+	bytes = append(bytes, []byte(b.Name)...)
+	return
+}
+
+// BootImageFromBytes returns a deserialized BootImage struct from bytes as well
+// as the number of bytes read from the slice.
+func bootImageFromBytes(bytes []byte) (*BootImage, int, error) {
+	// If less than length of boot image ID and count, it's probably invalid.
+	if len(bytes) < 5 {
+		return nil, 0, fmt.Errorf("not enough bytes for BootImage")
+	}
+	imageID := bootImageIDFromBytes(bytes[:4])
+	nameLength := int(bytes[4])
+	if 5+nameLength > len(bytes) {
+		return nil, 0, fmt.Errorf("not enough bytes for BootImage")
+	}
+	name := string(bytes[5 : 5+nameLength])
+	return &BootImage{ID: imageID, Name: name}, 5 + nameLength, nil
+}
+
+// makeVendorClassIdentifier calls the sysctl syscall on macOS to get the
+// platform model.
+func makeVendorClassIdentifier() (string, error) {
+	// Fetch hardware model for class ID.
+	hwModel, err := syscall.Sysctl("hw.model")
+	if err != nil {
+		return "", err
+	}
+	vendorClassID := fmt.Sprintf("AAPLBSDPC/i386/%s", hwModel)
+	return vendorClassID, nil
+}
+
+// parseBootImagesFromBSDPOption parses data from the BSDPOptionBootImageList
+// option and returns a list of BootImages.
+func parseBootImagesFromBSDPOption(data []byte) ([]BootImage, error) {
+	// Should at least have the # bytes of boot images.
+	if len(data) < 4 {
+		return nil, fmt.Errorf("invalid length boot image list")
+	}
+
+	readByteCount := 0
+	start := data
+	var bootImages []BootImage
+	for {
+		bootImage, readBytes, err := bootImageFromBytes(start)
+		if err != nil {
+			return nil, err
+		}
+		bootImages = append(bootImages, *bootImage)
+		readByteCount += readBytes
+		if readByteCount+1 >= len(data) {
+			break
+		}
+		start = start[readByteCount:]
+	}
+
+	return bootImages, nil
+}
+
+// parseVendorOptionsFromOptions extracts the sub-options list of the vendor-
+// specific options from the larger DHCP options list.
+func parseVendorOptionsFromOptions(options []Option) []Option {
+	var vendorOpts []Option
+	var err error
+	for _, opt := range options {
+		if opt.Code == OptionVendorSpecificInformation {
+			vendorOpts, err = OptionsFromBytes(opt.Data)
+			if err != nil {
+				return []Option{}
```

Comment:
I prefer the error not to be skipped silently, can you `log` it?

Reply:
I can, but there currently aren't any other usages of logging in the package. Should I use the global logger?

Reply:
yes, the global logger is OK here